File: src/main/java/org/entities/aggressive/ShortGolem.java
package org.entities.aggressive;

import org.engine.Main;
import org.entities.Autonomous;
import org.entities.Entity;
import org.entities.SmartRectangle;
import org.graphics.Animator;
import org.graphics.Graphics;
import org.level.Level;
import org.level.LevelController;
import org.loader.ImageResource;
import org.loader.ResourceHandler;
import org.world.Attack;
import org.world.HeightMap;
import org.world.HeightReturn;
import org.world.World;

//This class encompasses the blue, green, and red golems, as their classes are very similar

public class ShortGolem extends Autonomous {
    public static final int BLUE=0,GREEN=1,RED=2,PURPLE=3;
    private int golemType;
    private Animator golemAnimator;
    private ImageResource golem;
    private SmartRectangle hitbox=new SmartRectangle(x,y,width,height);
    public ShortGolem(int golemType, int subLevel, float spawnX, float spawnY) {
        super(subLevel, spawnX, spawnY);
        this.golemType=golemType;
        switch(golemType){
            case BLUE:
                health=2;
                golemAnimator=new Animator(ResourceHandler.getGolemLoader().getBlueGolemWalk(direction),10);
                displayName="Blue Golem";
                break;
            case GREEN:
                health=2;
                golemAnimator=new Animator(ResourceHandler.getGolemLoader().getGreenGolemWalk(direction),10);
                displayName="Green Golem";
                break;
            case RED:
                health=3;
                golemAnimator=new Animator(ResourceHandler.getGolemLoader().getRedGolemWalk(direction),10);
                displayName="Red Golem";
                break;
            case PURPLE:
                health=2;
                golemAnimator=new Animator(ResourceHandler.getGolemLoader().getPurpleGolemWalk(direction),10);
                displayName="Purple Golem";
                break;
        }
    }

    @Override
    public void update() {
        HeightReturn h= HeightMap.onGround(hitbox);

        //Action input
        if(damageTakenFrame==0) {
            if (direction) {
                vX = .25f;
            } else {
                vX = -.25f;
            }
        }else{
            if((direction&&!attackerBehind)||(!direction&&attackerBehind))vX=-.7f;
            else vX=.7f;
            damageTakenFrame--;
        }

        //Calculations
        y+=vY;
        vY-= World.getGravity();

        //X-velocity stuff
        boolean doXCalc=true;

        if (HeightMap.checkRightCollision(hitbox)) {
            if (x + width + 0.5>= HeightMap.findApplicable(hitbox,true).getStartX()) {
                if (vX < 0) x += vX;
                else vX=0;
                doXCalc=false;
                direction=!direction;
                x-=.25f;
            }
        }
        if(HeightMap.checkLeftCollision(hitbox)){
            if(x-0.5<=HeightMap.findApplicable(hitbox,false).getEndX()){
                if(vX>0)x+=vX;
                else vX=0;
                doXCalc=false;
                direction=!direction;
                x+=.25f;
            }
        }

        if(doXCalc){
            x+=vX;
            doXCalc();
        }

        if(damageCooldown>0){
            damageCooldown--;
            state=2;
        }else if(state==2){
            state=0;
        }

        //Y-velocity and ground calc
        if(h.isOnGround()&&vY<0){
            y=h.getGroundLevel();
            vY=0;
        }
        doAttackCalc();

        if(vX==0) {
            switch(golemType) {
                case BLUE:
                    golem = ResourceHandler.getGolemLoader().getBlueGolem(direction);
                    break;
                case GREEN:
                    golem = ResourceHandler.getGolemLoader().getGreenGolem(direction);
                    break;
                case RED:
                    golem = ResourceHandler.getGolemLoader().getRedGolem(direction);
                    break;
                case PURPLE:
                    golem=ResourceHandler.getGolemLoader().getPurpleGolem(direction);
                    break;
            }
        }else{
            golem=golemAnimator.getCurrentFrame();
        }
        if(state==0){
            switch(golemType) {
                case BLUE:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getBlueGolemWalk(direction));
                    break;
                case GREEN:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getGreenGolemWalk(direction));
                    break;
                case RED:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getRedGolemWalk(direction));
                    break;
                case PURPLE:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getPurpleGolemWalk(direction));
                    break;
            }
        }else if(state==1){
            switch(golemType) {
                case BLUE:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getBlueGolemAttack(direction));
                    break;
                case GREEN:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getGreenGolemAttack(direction));
                    break;
                case RED:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getRedGolemAttack(direction));
                    break;
                case PURPLE:
                    golemAnimator.setFrames(ResourceHandler.getGolemLoader().getPurpleGolemAttack(direction));
                    break;
            }
        }else if(state==2){
            switch (golemType) {
                case BLUE:
                    golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getBlueGolemKnockback(direction)});
                    break;
                case GREEN:
                    golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getGreenGolemKnockback(direction)});
                    break;
                case RED:
                    golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getRedGolemKnockback(direction)});
                    break;
                case PURPLE:
                    golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getPurpleGolemKnockback(direction)});
                    break;
            }
        }
        golemAnimator.update();
        if(state==1&&golemAnimator.getCurrentFrameNum()==golemAnimator.getFrames().length-1){
            state=0;
            Attack.attack(this,1,4);
        }
    }

    private void doXCalc(){
        Level l= LevelController.getCurrentLevel();
        if(x<l.getLeftLimit()||x+width>l.getRightLimit())direction=!direction;
        if(golemType==GREEN&&HeightMap.onEdge(hitbox,direction))direction=!direction;
    }

    private void doAttackCalc(){
        if(attackCooldown>0){
            attackCooldown--;
            return;
        }
        if(Main.getHarold().getY()>y+height||Main.getHarold().getY()+Main.getHarold().getWidth()<y){
            state=0;
            return;
        }
        if(direction){
            if(Main.getHarold().getX()>=x&&Main.getHarold().getX()<=x+width+4){
                state=1;
                golemAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }else{
            if(Main.getHarold().getX()+Main.getHarold().getWidth()<=x+width&&Main.getHarold().getWidth()+Main.getHarold().getX()>=x-4){
                state=1;
                golemAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }
    }

    @Override
    public void render() {
        width= Graphics.convertToWorldWidth(golem.getTexture().getWidth());
        height=Graphics.convertToWorldHeight(golem.getTexture().getHeight());
        hitbox.updateBounds(x,y,width,height);
        if(damageTakenFrame>0){
            Graphics.setColor(1,0,0,1);//Set damage color if needed
        }
        else Graphics.setColor(1,1,1,1);
        Graphics.drawImage(golem,x,y);
    }

    @Override
    public void reset() {

    }

    @Override
    public String toString() {
        switch(golemType) {
            case BLUE:
                return "Blue Golem @ " + x + "," + y;
            case GREEN:
                return "Green Golem @ " + x + "," + y;
            case RED:
                return "Red Golem @ " + x + "," + y;
            case PURPLE:
                return "Purple Golem @ "+x+","+y;
            default:
                return "Golem @ " + x + "," + y;
        }
    }

    @Override
    public void doDamage(Entity attacker, int damage) {
        if(golemType!=PURPLE)super.doDamage(attacker, damage);
        else{
            if(direction)attackerBehind=attacker.getX()<x;
            else attackerBehind=attacker.getX()>x;
            if(attackerBehind)direction=!direction;
            else if(!invincible)super.doDamage(attacker,damage);
        }
    }
}

File: src/main/java/org/entities/aggressive/ShortGolem.kt
package org.entities.aggressive

import org.engine.Main
import org.entities.Autonomous
import org.entities.Entity
import org.entities.SmartRectangle
import org.graphics.Animator
import org.graphics.Graphics
import org.level.LevelController.currentLevel
import org.loader.ImageResource
import org.loader.ResourceHandler
import org.world.Attack.melee
import org.world.HeightMap
import org.world.World.gravity

//This class encompasses the blue, green, and red golems, as their classes are very similar
class ShortGolem(private val golemType: GolemColor, subLevel: Int, spawnX: Float, spawnY: Float) :
    Autonomous(subLevel, spawnX, spawnY) {
    private var golemAnimator: Animator? = null
    private var golem: ImageResource? = null
    private val hitbox = SmartRectangle(x, y, width, height)
    override fun update(deltaTime: Float) {
        val h = HeightMap.onGround(hitbox)

        //Action input
        if (damageTakenFrame == 0f) {
            vX = if (direction) {
                20f
            } else {
                -20f
            }
        } else {
            vX = if (direction && !attackerBehind || !direction && attackerBehind) -10f else 10f
            damageTakenFrame = maxOf(damageTakenFrame - 100 * deltaTime, 0f)
        }

        //Calculations
        y += vY * deltaTime
        vY -= gravity * deltaTime * 3

        //X-velocity stuff
        var doXCalc = true
        if (HeightMap.checkRightCollision(hitbox)) {
            val hv = HeightMap.findApplicable(hitbox, true)
            if (hv != null && x + width + 0.5 >= hv.startX) {
                if (vX < 0) x += vX * deltaTime else vX = 0f
                doXCalc = false
                direction = !direction
                x -= .25f
            }
        }
        if (HeightMap.checkLeftCollision(hitbox)) {
            val hv = HeightMap.findApplicable(hitbox, false)
            if (hv != null && x - 0.5 <= hv.endX) {
                if (vX > 0) x += vX * deltaTime else vX = 0f
                doXCalc = false
                direction = !direction
                x += .25f
            }
        }
        if (doXCalc) {
            x += vX * deltaTime
            doXCalc()
        }
        if (damageCooldown > 0) {
            damageCooldown = maxOf(damageCooldown - 100 * deltaTime, 0f)
            state = 2
        } else if (state == 2) {
            state = 0
        }

        //Y-velocity and ground calc
        if (h.isOnGround && vY < 0) {
            y = h.groundLevel
            vY = 0f
        }
        doAttackCalc(deltaTime)
        if (vX == 0f) {
            golem = when (golemType) {
                GolemColor.BLUE -> ResourceHandler.getGolemLoader().getBlueGolem(direction)
                GolemColor.GREEN -> ResourceHandler.getGolemLoader().getGreenGolem(direction)
                GolemColor.RED -> ResourceHandler.getGolemLoader().getRedGolem(direction)
                GolemColor.PURPLE -> ResourceHandler.getGolemLoader().getPurpleGolem(direction)
            }
        } else {
            doSpriteCalc()
        }
        golemAnimator!!.update()
        if (state == 1 && golemAnimator!!.onLastFrame()) {
            state = 0
            melee(this, 1, 4f)
        }
    }

    private fun doXCalc() {
        val l = currentLevel
        if (x < l.leftLimit || x + width > l.rightLimit) direction = !direction
        if (golemType == GolemColor.GREEN && HeightMap.onEdge(hitbox, direction)) direction = !direction
    }

    private fun doSpriteCalc() {
        golem = golemAnimator!!.currentFrame
        if (state == 0) {
            when (golemType) {
                GolemColor.BLUE -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getBlueGolemWalk(direction)
                GolemColor.GREEN -> golemAnimator!!.frames =
                    ResourceHandler.getGolemLoader().getGreenGolemWalk(direction)
                GolemColor.RED -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getRedGolemWalk(direction)
                GolemColor.PURPLE -> golemAnimator!!.frames =
                    ResourceHandler.getGolemLoader().getPurpleGolemWalk(direction)
            }
        } else if (state == 1) {
            when (golemType) {
                GolemColor.BLUE -> golemAnimator!!.frames =
                    ResourceHandler.getGolemLoader().getBlueGolemAttack(direction)
                GolemColor.GREEN -> golemAnimator!!.frames =
                    ResourceHandler.getGolemLoader().getGreenGolemAttack(direction)
                GolemColor.RED -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getRedGolemAttack(direction)
                GolemColor.PURPLE -> golemAnimator!!.frames =
                    ResourceHandler.getGolemLoader().getPurpleGolemAttack(direction)
            }
        } else if (state == 2) {
            when (golemType) {
                GolemColor.BLUE -> golemAnimator!!.frames =
                    arrayOf(ResourceHandler.getGolemLoader().getBlueGolemKnockback(direction))
                GolemColor.GREEN -> golemAnimator!!.frames =
                    arrayOf(ResourceHandler.getGolemLoader().getGreenGolemKnockback(direction))
                GolemColor.RED -> golemAnimator!!.frames =
                    arrayOf(ResourceHandler.getGolemLoader().getRedGolemKnockback(direction))
                GolemColor.PURPLE -> golemAnimator!!.frames =
                    arrayOf(ResourceHandler.getGolemLoader().getPurpleGolemKnockback(direction))
            }
        }
    }

    private fun doAttackCalc(deltaTime: Float) {
        if (attackCooldown > 0) {
            attackCooldown -= 100 * deltaTime
            return
        }
        if (Main.getHarold().y > y + height || Main.getHarold().y + Main.getHarold().width < y) {
            state = 0
            return
        }
        if (direction) {
            if (Main.getHarold().x >= x && Main.getHarold().x <= x + width + 4) {
                state = 1
                golemAnimator!!.setCurrentFrame(0)
                attackCooldown = 80f
            }
        } else {
            if (Main.getHarold().x + Main.getHarold().width <= x + width && Main.getHarold().width + Main.getHarold().x >= x - 4) {
                state = 1
                golemAnimator!!.setCurrentFrame(0)
                attackCooldown = 80f
            }
        }
    }

    override fun doDamage(attacker: Entity, damage: Int) {
        if (golemType != GolemColor.PURPLE) super.doDamage(attacker, damage) else {
            attackerBehind = if (direction) attacker.x < x else attacker.x > x
            if (attackerBehind) direction = !direction else if (!invincible) super.doDamage(attacker, damage)
        }
    }

    override fun render() {
        if (golem == null) return
        width = Graphics.toWorldWidth(golem!!.texture.width.toFloat())
        height = Graphics.toWorldHeight(golem!!.texture.height.toFloat())
        hitbox.updateBounds(x, y, width, height)
        if (damageTakenFrame > 0) {
            Graphics.setDrawColor(1f, 0f, 0f, 1f) //Set damage color if needed
        } else {
            Graphics.setDrawColor(1f, 1f, 1f, 1f)
        }
        Graphics.drawImage(golem!!, x, y)
        Graphics.setDrawColor(1f, 1f, 1f, 1f)
    }

    override fun reset() {}
    override fun toString(): String {
        return when (golemType) {
            GolemColor.BLUE -> "Blue Golem @ $x,$y"
            GolemColor.GREEN -> "Green Golem @ $x,$y"
            GolemColor.RED -> "Red Golem @ $x,$y"
            GolemColor.PURPLE -> "Purple Golem @ $x,$y"
            else -> super.toString()
        }
    }

    enum class GolemColor {
        BLUE, GREEN, RED, PURPLE
    }

    init {
        when (golemType) {
            GolemColor.BLUE -> {
                health = 2
                golemAnimator = Animator(ResourceHandler.getGolemLoader().getBlueGolemWalk(direction), 10)
                displayName = "Blue Golem"
            }
            GolemColor.GREEN -> {
                health = 2
                golemAnimator = Animator(ResourceHandler.getGolemLoader().getGreenGolemWalk(direction), 10)
                displayName = "Green Golem"
            }
            GolemColor.RED -> {
                health = 3
                golemAnimator = Animator(ResourceHandler.getGolemLoader().getRedGolemWalk(direction), 10)
                displayName = "Red Golem"
            }
            GolemColor.PURPLE -> {
                health = 2
                golemAnimator = Animator(ResourceHandler.getGolemLoader().getPurpleGolemWalk(direction), 10)
                displayName = "Purple Golem"
            }
        }
    }
}